cpp/ql/test/query-tests/Security/CWE/CWE-367/semmle/test2.cpp

```diff
@@ -0,0 +1,333 @@
+		// retry
+		f = fopen(path, "r"); // GOOD (this is just trying again) [FALSE POSITIVE]
+	}
+
+	// ...
+}
+
+void test1_2(const char *path)
+{
+	FILE *f = NULL;
+
+	// try until we succeed
+	while (f == NULL)
+	{
+		f = fopen(path, "r"); // GOOD (this is just trying again) [FALSE POSITIVE]
+
+		// ...
+	}
+
+	// ...
+}
+
+// --- stat -> open ---
+
+void test2_1(const char *path)
+{
+	FILE *f = NULL;
+	stat_data buf;
+
+	if (stat(path, &buf))
+	{
+		f = fopen(path, "r"); // BAD
+	}
+
+	// ...
+}
+
+void test2_2(const char *path)
+{
+	FILE *f = NULL;
+	stat_data buf;
+
+	stat(path, &buf);
+	if (buf.foo > 0)
+	{
+		f = fopen(path, "r"); // BAD [NOT DETECTED]
+	}
+
+	// ...
+}
+
+void test2_3(const char *path)
+{
+	FILE *f = NULL;
+	stat_data buf;
+	stat_data *buf_ptr = &buf;
+
+	stat(path, buf_ptr);
+	if (buf_ptr->foo > 0)
+	{
+		f = fopen(path, "r"); // BAD
+	}
+
+	// ...
+}
+
+bool stat_condition(const stat_data *buf);
+bool other_condition();
+
+void test2_4(const char *path)
+{
+	FILE *f = NULL;
+	stat_data buf;
+
+	stat(path, &buf);
+	if (stat_condition(&buf))
+	{
+		f = fopen(path, "r"); // BAD [NOT DETECTED]
+	}
+
+	// ...
+}
+
+void test2_5(const char *path)
+{
+	FILE *f = NULL;
+	stat_data buf;
+	stat_data *buf_ptr = &buf;
+
+	stat(path, buf_ptr);
+	if (stat_condition(buf_ptr))
+	{
+		f = fopen(path, "r"); // BAD [NOT DETECTED]
+	}
+
+	// ...
+}
+
+void test2_6(const char *path)
+{
+	FILE *f = NULL;
+	stat_data buf;
+
+	stat(path, &buf);
+	if (other_condition())
+	{
+		f = fopen(path, "r"); // GOOD (does not depend on the result of stat)
+	}
+
+	// ...
+}
+
+void test2_7(const char *path, int arg)
+{
+	stat_data buf;
+	int f;
+
+	if (stat(path, &buf))
+	{
+		f = open(path, arg); // BAD
+	}
+
+	// ...
+}
+
+void test2_8(const char *path, int arg)
+{
+	stat_data buf;
+	int f;
+
+	if (lstat(path, &buf))
+	{
+		f = open(path, arg); // BAD
+	}
+
+	// ...
+}
+
+void test2_9(const char *path, int arg)
+{
+	stat_data buf;
+	int f;
+
+	if (stat(path, &buf))
+	{
+		f = creat(path, arg); // BAD [NOT DETECTED]
+	}
+
+	// ...
+}
+
+void test2_10(int dir, const char *path, int arg)
+{
+	stat_data buf;
+	int f;
+
+	if (fstatat(dir, path, &buf))
+	{
+		f = openat(dir, path, arg); // BAD [NOT DETECTED]
+	}
+
+	// ...
+}
+
+// --- open -> stat ---
+
+void test3_1(const char *path, int arg)
+{
+	stat_data buf;
+	int f;
+
+	f = open(path, arg);
+	if (stat(path, &buf)) // BAD [NOT DETECTED]
+	{
+		// ...
+	}
+
+	// ...
+}
+
+void test3_2(const char *path, int arg)
+{
+	stat_data buf;
+	int f;
+
+	f = open(path, arg);
+	if (fstat(f, &buf)) // GOOD (uses file descriptor, not path)
+	{
+		// ...
+	}
+
+	// ...
+}
+
+// --- open -> chmod ---
+
+void test4_1(const char *path)
+{
+	FILE *f = NULL;
+
+	f = fopen(path, "w");
+	if (f)
+	{
+		// ...
+
+		fclose(f);
+
+		chmod(path, 0); // DUBIOUS (bad but perhaps not exploitable) [REPORTED]
+	}
+}
+
+// --- rename -> remove / open ---
+
+void test5_1(const char *path1, const char *path2)
+{
+	if (!rename(path1, path2))
```

Comment:
I think these cases are swapped from what we'd see in real code, since `rename` returns 0 on success

Reply:
Well spotted.  Fixed.